Unified paging

jeromekelleher opened the pull request:
This is an early PR for comments and discussion. 

Here is a first pass at implementing paging in a general way. The idea is that the `Backend` object gets a call to `searchX` and then delegates this to `runRequest`, which does the paging in a general way. `runRequest` gets a bunch of parameters which allow it to do this, the key one being `objectGenerator`. This is an iterator over all of the objects specified by the request.

The tricky bit is dealing with page tokens. To deal with this, the generator yields pairs of values `(object, nextPageToken)`, which it's the generators responsibility to manage. It's easy in this case, but it'll be more trick when I move on to variants.

Before doing this, I'd like to get some feedback. What do people think of the general outline?

pashields commented on ga4gh/backends.py:
```diff
@@ -415,35 +413,50 @@ def searchVariants(self, request):
+        pageToken = None
+        for obj, pageToken in objectGenerator(request):
+            pageList.append(obj)
+            if len(pageList) == request.pageSize:
```
>= would save us from all sorts of slippery cases (pageSize is 0, None, or negative for instance).

jeromekelleher replied:
Good call @pashields, thanks.

jeromekelleher commented:
This is now ready for discussion and (hopefully!) merging.

I've implemented unified paging, as well as doing a much-needed refactoring on some of the more important classes. Here's a summary:

- Paging is now done centrally on the backend object. This is now seen as the central point for the Flask frontend to send all requests. We get the objects from generators for the objects in question; these generators must return `(object, nextPageToken)` pairs and be able to resume iteration from any point.
- The Backend object has been refactored so that there is only one class (we no longer have the `WormtableBackend` and `TabixBackend` classes). We set a variable on the Flask app (`app.backend`) to allow the calls to get routed to the correct place.
- Various minor tweaks and PEP8 updates.
- The `Backend` API has now been changed so that it takes text versions of the requests and returns text versions of the responses. This simplifies the frontend substantially, and allows us a lot of flexibility in how we generate the string responses.

There is quite a lot in here, and it's probably too big for a single PR. However, I think the cumulative effect of these changes is a codebase that's substantially easier to understand and work with, so I hope it's not too much hassle reviewing it. Apologies also if I've clobbered any one else's work with this update.

kerrydc commented on ga4gh/backends.py:
```diff
@@ -368,99 +311,104 @@ def getMetadata(self):
+        (object, nextPageToken) pairs, and be able to resume iteration from
+        any point using the nextPageToken attribute of the request object.
+        """
+        request = requestClass.fromJSON(requestStr)
```
I would call validate here, and therefore either add a validate JSON string method to protocol or else pass in the JSON dict here.

jeromekelleher replied:
Agreed.  The required method is being added in #64, so I've put in a TODO.

kerrydc commented on ga4gh/backends.py:
```diff
@@ -368,99 +311,104 @@ def getMetadata(self):
+    def searchVariants(self, request):
+        """
+        Returns a GASearchVariantsResponse for the specified
+        GASearchVariantsRequest  object.
```
nit: remove one space.

jeromekelleher replied:
From where? Sorry, I'm not seeing it.

kerrydc replied:
GASearchVariantsRequest  object.
_____________________^^_____

jeromekelleher replied:
Aha, spotted it now! Thanks, was being a bit blind there.

kerrydc commented on ga4gh/backends.py:
```diff
@@ -368,99 +311,104 @@ def getMetadata(self):
+        """
         currentIndex = 0
         if request.pageToken is not None:
             currentIndex = int(request.pageToken)
```
This is assuming that pageToken can be parsed as an int. That is not always the case.

jeromekelleher replied:
It is in this case though. PageTokens are handled locally here, so that each generator is locally responsible for managing its own. So, for variantSets, the pageToken is always an int, and for variants, the pageToken is always a ":" seperated pair of ints. We'll have to do something a bit more complicated with these when we start supporting Datasets, but I think the extension should be easy enough.

bioinformed replied:
Not trying to rush anyone, but as with the previous round of cleanups, I'm
waiting for this refactoring to land before I implement the new htslib
VCF/BCF backend.

On Mon, Jan 12, 2015 at 7:32 AM, Jerome Kelleher <notifications@github.com>
wrote:

> In ga4gh/backends.py
> <https://github.com/ga4gh/server/pull/59#discussion-diff-22796944>:
>
> > +
> > +    def searchVariants(self, request):
> > +        """
> > +        Returns a GASearchVariantsResponse for the specified
> > +        GASearchVariantsRequest  object.
> > +        """
> > +        return self.runSearchRequest(
> > +            request, protocol.GASearchVariantsRequest,
> > +            protocol.GASearchVariantsResponse, "variants",
> > +            self.variantsGenerator)
> > +
> > +    def variantSetsGenerator(self, request):
> > +        """
> > +        Returns a generator over the (variantSet, nextPageToken) pairs defined
> > +        by the speficied request.
> > +        """
> >          currentIndex = 0
> >          if request.pageToken is not None:
> >              currentIndex = int(request.pageToken)
>
> It is in this case though. PageTokens are handled locally here, so that
> each generator is locally responsible for managing its own. So, for
> variantSets, the pageToken is always an int, and for variants, the
> pageToken is always a ":" seperated pair of ints. We'll have to do
> something a bit more complicated with these when we start supporting
> Datasets, but I think the extension should be easy enough.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ga4gh/server/pull/59/files#r22796944>.
>

kerrydc commented on ga4gh/backends.py:
```diff
@@ -368,99 +311,104 @@ def getMetadata(self):
+        startPosition = request.start
+        if request.pageToken is not None:
+            startVariantSetIndex, startPosition = map(
+                int, request.pageToken.split(':'))
```
If we assume pageToken has some structure (str containing ':', for example) we should validate it and throw an error if it does not conform.

jeromekelleher replied:
This is a good idea. I've added a new method `parsePageToken` that verifies and parses the pageToken. It might work out that assuming all parts of the token must be integers is restrictive, but I think it's worth trying to do this for a start. It's a lot harder for an attacker to do nasty injection attacks when they are restricted to using integers.

kerrydc commented on ga4gh/server/__init__.py:
```diff
@@ -6,6 +10,6 @@
 # TODO: Replace hard-coded value with a flag or conifg setting.
-app.config['MAX_CONTENT_LENGTH'] = 2 * 1024 * 1024 # 2MB 
+app.config['MAX_CONTENT_LENGTH'] = 2 * 1024 * 1024  # 2MB
 
```
I think you pulled this in by accident. Please remove.

pashields replied:
This looks a pep8 fix @kerrydc. It should be covered by #63 though, so it will just get fixed in the a merge or rebase.

jeromekelleher replied:
This was convergent evolution all right! It'll get fixed when I rebase.

kerrydc commented on tests/test_backends.py:
```diff
@@ -117,8 +117,8 @@ def resultIterator(
-            self.assertEqual(type(response), ResponseClass)
+            responseStr = searchMethod(request.toJSON())
+            response = ResponseClass.fromJSON(responseStr)
             l = getattr(response, listMember)
```
Validate the response?

jeromekelleher replied:
Definitely. Again, the required method is being added in #64, so I've put in a TODO.

kerrydc commented:
Mostly looks good, but it needs a lot of unit test coverage. I would suggest that your test cases cover both legal and illegal input, and that you exercise raising all errors that can be raised.

kerrydc commented:
Mostly looks good, but it needs a lot of unit test coverage. I would suggest that your test cases cover both legal and illegal input, and that you exercise raising all errors that can be raised.

kerrydc commented:
Also, perhaps while writing tests you could update the style and variable names in test_backends.py. I'm working in there as well, adding Tabix backend tests, so please don't bother adding those.

jeromekelleher commented:
Thanks for the feedback @kerrydc, it's been very useful. I've implemented the majority of them, and updated the style on the variant tests. I haven't added any tests on failure conditions, as I don't see much point in this until we define a proper error model (i.e., we define the exceptions we expect and when we should see them; this should be something we tackle quite soon I think). I also think that the test cases need a decent refactoring, and it would be best to wait until you've finished working on the TabixBackend tests before this is undertaken. 

Also, in the interest of not holding @bioinformed up any longer, it would be good to try to wrap this up and vote on it.

kerrydc commented on ga4gh/backends.py:
```diff
@@ -368,99 +311,135 @@ def getMetadata(self):
+        not conform to this specification, raise a InvalidPageToken
+        exception.
+        """
+        tokens = pageToken.split(":")
```
What happens when pageToken is an int or None?

jeromekelleher replied:
Well, `nextPageToken` is a string in the protocol, so we can safely assume that conforms to this (once we start calling `validate()` on our input request). We've already tested that it's not `None` when we call this method.

We're changing the way things are done slightly here, as pageTokens will not be exactly the same form as before. I think this is a lot simpler though, and should generalise nicely to working with dataSets too. We do have reasonably good test coverage on the paging code, so I'm fairly sure it works properly in the nominal case. We definitely need to write some test code to break it though, and deal with these error cases properly.

Naburimannu commented:
I may be missing something, but it looks like this also solves #10? I can't spot the process ever getting larger than ~210 MB.

jeromekelleher commented:
It shouldn't have made any real difference to #10 - we're still building up the list of all objects and returning them. Perhaps we've change the default `pageSize` or something? I think we had some extra code in there to set it before?

kerrydc commented:
+1, thanks for the test fixes. We should add more test cases, but that can wait for a future PR.

kerrydc commented:
Jerome, anything outstanding on this PR? If not, let us vote! I'll merge on another two +1s.

Naburimannu commented:
Performance / memory footprint has definitely changed since I did my initial measurements for #10, but I verified this morning that it is also changed at tip-of-tree, so something else must be at fault.

+1

pashields commented:
+1

jeromekelleher commented:
I'm happy to merge this ASAP. Let's follow up with more test cases in a later PR.
